### PR TITLE
Committer Policy

### PR DESCRIPTION
**Pull request by paulidale**
Based over #6 to share the definitions.

Aiming to replace: https://www.openssl.org/policies/committers.html

### CODE REVIEW

**Comment by paulidale** [comment]
Markdown conversion and basic import of the current policy (with a few minor corrections).
Links and other updates to occur in due course.

File: policies/committer-policy.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+## Approvals and code reviews
+
+All submissions must be reviewed and approved by at least two committers,
+one of whom must also be an OTC member.  If the author is also a committer
+then that counts as one of the reviews.  In other words:
+
+* OTC members need one approval from any committer
+* Committers need one approval from an OTC member
+* Contributors without commit rights need two approvals, including
```
Comment (paulidale) [inline policies/committer-policy.md:45]: I'm having some concerns about author's reviews not counting for them selves.

* releases require commits to be reviewed -- does this mean three people for each release instead of two?  If so, who?  We have trouble finding two at times.  These commits are mechanical and a second set of eyes seems reasonable.
* If a third party submits a PR and one of us adds a test case, does this make them unable to review any of the submission?  The submission would require two additional reviews for the test case but the entire thing?
* If two of us cooperate to create a PR (which happens more than occasionally), does this require two new reviewers?

There are four staff who can currently sign off reviews and by far the bulk of reviewing is done by these four.  Insisting all four sign off seems problematic, even three could be difficult.

**Comment by paulidale** [comment]
Still some more definitions/cross referencing to be done.

File: policies/committer-policy.md (hunk 1 above)
Comment (t-j-h) [reply policies/committer-policy.md:45]: I think we can put in place a policy that handles the mechanical items needing during a release as a separate context where basically the two people involved in the release are the reviewers - as basically it is mechanical stuff rather than creative and in need of dual review.

For test case addition - they can always be placed in a separate PR if you are concerned about review challenges.

And as per previous discussions, we can make a policy change like this and see how it works out in practice and adjust after we have more information. Not changing a policy because we are concerned about something that isn't certain to happen is concerning.

And of course the number of paid resources is increasing so the "four staff" issue should be less of a concern over time.

File: policies/committer-policy.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  one from an OTC member.
+
+An OMC member may apply an OMC-hold to a submission.  An OTC member may
```
Comment (t-j-h) [inline policies/committer-policy.md:48]: This needs to be more clearly worded in terms of use of github labels - as that is what we are using.

File: policies/committer-policy.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+are nevertheless welcome to comment on submissions and do technical
+reviews.  We always value another pair of eyes, and volunteering for
+reviews counts favourably towards becoming a committer.  As an author,
+we ask that you address all comments, even if you already have the
+necessary approvals.
+
+If you have trouble finding consensus on a difficult review, reach out to
+the OTC at `otc@openssl.org` (private, moderated) or the project
+at `openssl-project@openssl.org` (public, moderated).  On GitHub, you can
+target the OMC members with `@openssl/omc`, OTC members with `@openssl/otc`,
```
Comment (t-j-h) [inline policies/committer-policy.md:68]: "target" is a poor choice of word - "reach" would fit the context better.

File: policies/committer-policy.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+security issue, we ask that you report it to `openssl-security@openssl.org`
+and don't discuss it further in public.  We review security issues
+privately, however acceptance of a submission for a security issue does
+not bypass the review process that applies to all submissions.
+
+## A note on CLAs
+
+All authors, including committers, must have current CLAs on file.  A CLA
+is not required for trivial contributions (e.g. the fix of a spelling
+mistake).  Refer to the CLA page for further details.
```
Comment (t-j-h) [inline policies/committer-policy.md:92]: "CLA page" should be a link to the details - the user shouldn't have to go hunting for a page - direct reference.

File: policies/committer-policy.md (hunk 1 above)
Comment (paulidale) [reply policies/committer-policy.md:45]: Paid resources don't being as either committer or OTC.  Thus, they don't have any official review capability.  At this stage, we don't know how long it takes.  Given the usual rate of progress, months seems likely.

I have worded out an exception for the release.  Being mechanical, there isn't much to review really.

I was struggling with the myriad other cases as exceptions.  I'll keep trying.  Personally, I don't have a problem with one submitter reviewing bits done by another so long as there is a third sign off.  Codifying this nicely, on the other hand, isn't so easy :)

Comment (kaduk) [reply policies/committer-policy.md:45]: If we are going to be talking about the author's review *not* counting as one of the two reviews, should the text in the PR be modified to reflect that?  I currently see

> If the author is also a committer then that counts as one of the reviews

which does not seem consistent with the ongoing discussion here.

File: policies/committer-policy.md (hunk 4 above)
Comment (paulidale) [reply policies/committer-policy.md:92]: Yes, more links required.  I realise I pushed without including CLA links.

File: policies/committer-policy.md (hunk 3 above)
Comment (paulidale) [reply policies/committer-policy.md:68]: Agreed, target is the original terminology.  I've still to clean this up.

File: policies/committer-policy.md (hunk 2 above)
Comment (paulidale) [reply policies/committer-policy.md]: I've made an attempt at this.  I'm not completely happy with it.

File: policies/committer-policy.md (hunk 1 above)
Comment (paulidale) [reply policies/committer-policy.md:45]: This is the original text slightly updated.

I'm torn between getting the current policy in and then updating it or trying to do the update as part of this PR.

Comment (paulidale) [reply policies/committer-policy.md:45]: How about this wording:

> All submissions must be reviewed and approved by at least two committers,
> one of whom must also be an OTC member.  Neither of the reviewers can
> be the author of the submission.
> 
> The sole exception to this is for the release procedure commits where
> the author's review does count towards the two needed.
> 
> In the case where two committers make a joint submission, they can review
> each other's code but not their own.  A third reviewer will be required.

To replace the text in the paragraph and the bullet points.

Comment (t-j-h) [reply policies/committer-policy.md:45]: That looks like a reasonable and pragmatic description to match the intent of the change.
Updating the policy is the path you've started down rather than just reformatting so I'd stick to that now.

Comment (paulidale) [reply policies/committer-policy.md:45]: Okay.  I'll still add this as a separate commit in case we need to revert back to what was there.

**Comment by paulidale** [comment]
Feedback addressed.

File: policies/committer-policy.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+## Who is a committer?
+
+OpenSSL committers are contributors who have commit access to the OpenSSL
+source code repository.  Committers review and commit their own patches
+as well as those of other contributors.
+
+## How to become a committer?
+
+Commit access is granted by the [OMC] typically on the recommendation
+of the [OTC] (see the [OpenSSL Bylaws]).
```
Comment (mattcaswell) [inline policies/committer-policy.md:13]: This is slightly confusing. The bylaws say:

"Commit access is granted by invitation from the OTC and requires a prior OMC vote of acceptance."

So the bylaws say "Commit access is granted by...the OTC", but this policy says its granted by the OMC.

File: policies/committer-policy.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+who contribute to all areas of the codebase.  All committers share
+the responsibility for the overall health of the project: aside from
+contributing quality features, committers are team players who fix bugs,
+address open issues, review community contributions, and improve tests
+and documentation.  Committers are also shepherds of the OpenSSL community
+and its code of conduct.
+
+To become a committer, start by contributing code.  Read our coding style,
+and get to know our build and test system.  Then, use the Github issue
+tracker, and our mailing lists find impactful ideas to work on.
```
Comment (mattcaswell) [inline policies/committer-policy.md:26]: to find?

File: policies/committer-policy.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+the project objectives or are otherwise disruptive, committer status
+may also be revoked by the OMC.
+
+## Approvals and code reviews
+
+All submissions must be reviewed and approved by at least two committers,
+one of whom must also be an OTC member.  Neither of the reviewers can
+be the author of the submission.
+
+The sole exception to this is for the release procedure commits where
```
Comment (mattcaswell) [inline policies/committer-policy.md:43]: There are different commits involved in the release. There are those automatically generated ones which come from the scripts. Also we often end up doing things like updating CHANGES/NEWS etc. We will have to get better at doing those ones earlier. I do worry that there will inevitably things that crop up during the release that we have to fix at the last moment - but I guess we'll have to see how that works out in practice. Should this statement in the policy be more explicit about referring to the automatically generated release procedure commits?

File: policies/committer-policy.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+the author's review does count towards the two needed.
+
+In the case where two committers make a joint submission, they can review
+each other's code but not their own.  A third reviewer will be required.
```
Comment (mattcaswell) [inline policies/committer-policy.md:47]: So, to be clear, does this mean that in the case where a single PR contains commits from 2 different committers, we only need one additional reviewer? The 2 authors each review the commits authored by the other person, and a 3rd committer reviews all of them?

File: policies/committer-policy.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+An OMC member may apply a _hold: needs OMC decision_ label to a submission.
+An OTC member may apply a _hold: needs OTC decision_ to a submission.
+A _hold: needs OMC decision_ label may removed by the member that put
```
Comment (mattcaswell) [inline policies/committer-policy.md:51]: be removed

File: policies/committer-policy.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+in place the hold or by a vote of the OMC.
```
Comment (mattcaswell) [inline policies/committer-policy.md:52]: perhaps "by a decision of the OMC"? I'm not sure a vote is always necessary if there is broad consensus

File: policies/committer-policy.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+A _hold: needs OTC decision_ label may be removed by the member that put
+in place the hold or by a vote of the OTC.
```
Comment (mattcaswell) [inline policies/committer-policy.md:54]: as above

File: policies/committer-policy.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+Approved submissions (outside of the automated release process and [NEWS]
+and [CHANGES] file updates) shall only be applied after a 24-hour delay from
+the approval (except for minor build and test breakage fix approvals).
```
Comment (mattcaswell) [inline policies/committer-policy.md:58]: I'd be tempted to drop "minor" here. Also should we mention the "urgent" label?

File: policies/committer-policy.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+If at any point during development or review you discover a potential
+security issue, we ask that you report it to `openssl-security@openssl.org`
+and don't discuss it further in public.  We review security issues
+privately, however acceptance of a submission for a security issue does
+not bypass the review process that applies to all submissions.
+
+## A note on CLAs
+
+All authors, including committers, must have current [CLA] on file.  A CLA
```
Comment (mattcaswell) [inline policies/committer-policy.md:92]: a current? The original policy says CLAs (plural) which is the case where a CCLA and ICLA are required.

File: policies/committer-policy.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+is not required for trivial contributions (e.g. the fix of a spelling
+mistake).  Refer to the [Contributor Agreements] page for further details.
```
Comment (mattcaswell) [inline policies/committer-policy.md:94]: Should we document the fact that all reviewers and the author should agree that a submission is trivial. This is what we do in practice so it should be in a formal policy somewhere.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: policies/committer-policy.md (hunk 5 above)
Comment (paulidale) [reply policies/committer-policy.md:13]: It's the OMC voting acceptance, they are granting.
The OTC recommends.

File: policies/committer-policy.md (hunk 7 above)
Comment (paulidale) [reply policies/committer-policy.md:43]: I omitted the automatically generated part on purpose.  I think it will cause more pain than it is worth.
I've no problem with adding it, it's someone else's pain :wink: 

We could make an attempt to track the manual changes to see how impactful it would be and add that later.

File: policies/committer-policy.md (hunk 8 above)
Comment (paulidale) [reply policies/committer-policy.md:47]: Yes.  Or a 4th also reviews everything (in which case there is no issue at all).

I had trouble capturing this intent without creating a word jam of undue complexity.

File: policies/committer-policy.md (hunk 10 above)
Comment (paulidale) [reply policies/committer-policy.md:52]: Agreed, although the bylaws technically allow no way to make a decision without a vote.

File: policies/committer-policy.md (hunk 11 above)
Comment (paulidale) [reply policies/committer-policy.md:54]: I'm tempted to make this: _by a decision of the OMC or OTC_.

File: policies/committer-policy.md (hunk 12 above)
Comment (paulidale) [reply policies/committer-policy.md:58]: This section mentions automated release process plus NEWS and CHANGES explicitly.  It should be bought in line with the earlier section.

Comment (paulidale) [reply policies/committer-policy.md:58]: I've mentioned the urgent label and removed _minor_.

File: policies/committer-policy.md (hunk 11 above)
Comment (mattcaswell) [reply policies/committer-policy.md]: I wouldn't mention OMC here. While strictly speaking the OMC is the ultimate authority and *could* do anything - that would be "by exception". By policy I wouldn't expect the OMC to mess with OTC labels.

File: policies/committer-policy.md (hunk 7 above)
Comment (mattcaswell) [reply policies/committer-policy.md:43]: I think we should be explicit about the automatically generated stuff. The alternative is that we allow a broader range of changes while doing the release procedure - but then I think it gets a lot more difficult to nail down exactly what is and is not allowed. The auto generated stuff is at least easy to define.

Comment (paulidale) [reply policies/committer-policy.md:43]: I've changed this to the automated changes plus NEWS and CHANGES edits.

File: policies/committer-policy.md (hunk 8 above)
Comment (t-j-h) [reply policies/committer-policy.md:51]: I'm not sure that this exception actually makes sense in practical terms. It would require a clearer view of who does what and a separate set of reviews for each commit and in practical terms that doesn't really seem to happen in PRs in general. I would leave this whole concept out and always have two reviewers who are not the author(s) and see how that works out in practical terms. i.e. this is a premature optimisation for a problem we haven't yet experienced with the changed review rule. We should IMHO put in the simple rule and adjust based on experience with the new rule over time.

Comment (paulidale) [reply policies/committer-policy.md:51]: I'm in total agreement that wording isn't ideal as it currently stands.

I also believe that two completely independent reviewers will be problematic in pretty much any instance where this clause applies.

File: policies/committer-policy.md [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+and its code of conduct.
+
+To become a committer, start by contributing code.  Read our coding style,
+and get to know our build and test system.  Then, use the Github issue
+tracker, and our mailing lists to find impactful ideas to work on.
+
+## How to maintain committer status?
+
+To maintain committer status, you must stay active in the project.  Refer
+to the OpenSSL Bylaws for details.
```
Comment (kroeckx) [inline policies/committer-policy.md:31]: And a link to the bylaws?

File: policies/committer-policy.md [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+All submissions must be reviewed and approved by at least two committers,
+one of whom must also be an OTC member.  Neither of the reviewers can
+be the author of the submission.
+
+The sole exception to this is for the release procedure commits where
+the author's review does count towards the two needed.
+
+The sole exception to this is during the release process where the
+author's review does count towards the two needed for the automated
+release process and [NEWS] and [CHANGES] file updates.
```
Comment (kroeckx) [inline policies/committer-policy.md:48]: The paragraph above this one should probably get removed for being duplicate.

File: policies/committer-policy.md (hunk 8 above)
Comment (kroeckx) [reply policies/committer-policy.md:51]: I don't actually see this changing anything, 3 people are needed for the normal case and this case.

File: policies/committer-policy.md (hunk 15 above)
Comment (paulidale) [reply policies/committer-policy.md:31]: Yep, via a glossary entry.

File: policies/committer-policy.md (hunk 16 above)
Comment (paulidale) [reply policies/committer-policy.md:48]: Thanks, missed this.

**Comment by paulidale** [comment]
Unless Kurt has further objections, I'd like to call a vote on this:
```
Topic: Accept the committer policy as of 3766d6ba2648e716e973af6e1821687ad46ee57c
Proposed by: Pauli
Issue link: https://github.com/openssl/general-policies/issues/8
Public: yes
Opened: 2022-02-17
Closed: YYYY-MM-DD
Accepted:  yes/no  (for: X, against: Y, abstained: Z, not voted: W)

  Kurt       [  ]
  Mark       [  ]
  Matt       [  ]
  Pauli      [+1]
  Richard    [  ]
  Tim        [  ]
```

**Comment by paulidale** [comment]
Paul: [+1]

**Comment by t-j-h** [comment]
Vote: [+1]

File: policies/committer-policy.md (hunk 8 above)
Comment (t8m) [reply policies/committer-policy.md:48]: I'm with @paulidale that to require 2 additional people to review something done by joint work of 2 committers would be seriously problematic. And I do not think the current wording is unclear or anything.

**Comment by t8m** [comment]
The bigger number of formal reviews will be required by the process the more formal and less thorough those reviews will be. IMO that is inevitable. However I think this change in requirements to always have two independent reviewers of a code someone wrote is still well within reasonable limits. I would just not expect this change to have as big impact on the code quality as perhaps someone might expect.

File: policies/committer-policy.md [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+In the unlikely and unfortunate event that your actions conflict with
+the project objectives or are otherwise disruptive, committer status
+may also be revoked by the OMC.
+
+## Approvals and code reviews
+
+All submissions must be reviewed and approved by at least two committers,
+one of whom must also be an OTC member.  Neither of the reviewers can
+be the author of the submission.
```
Comment (mspncp) [inline policies/committer-policy.md:41]: As a non-OMC member I ask myself: What is the rationale behind this significant tightening of the rules? 🤔

Comment (beldmit) [reply policies/committer-policy.md:41]: +1.

Comment (paulidale) [reply policies/committer-policy.md:41]: Mostly it represents a slackness in the current review process.

Comment (beldmit) [reply policies/committer-policy.md:41]: Well... I'm afraid it will not add much energy to it.

Comment (mattcaswell) [reply policies/committer-policy.md:41]: It's about improving quality and ensuring enough eyes look at every change.

File: policies/committer-policy.md [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+The sole exception to this is during the release process where the
+author's review does count towards the two needed for the automated
+release process and [NEWS] and [CHANGES] file updates.
+
+In the case where two committers make a joint submission, they can review
+each other's code but not their own.  A third reviewer will be required.
+
+An OMC member may apply a _hold: needs OMC decision_ label to a submission.
+An OTC member may apply a _hold: needs OTC decision_ to a submission.
```
Comment (mattcaswell) [inline policies/committer-policy.md:51]: Didn't think of this until now (so probably something to think about after the vote) - but do we mean to restrict things so that only OMC can place an OMC hold and only OTC can place an OTC hold? Shouldn't it be possible for any committer to refer a decision to OMC or OTC if they think it is appropriate?

File: policies/committer-policy.md (hunk 17 above)
Comment (beldmit) [reply policies/committer-policy.md:41]: I understand intentions but have doubts about results.

**Comment by mattcaswell** [comment]
> Accept the committer policy as of 3766d6ba2648e716e973af6e1821687ad46ee57c

Question: This PR seems to have a new unrelated commit in it to define perlasm. Its the same commit as is present in #10. The vote that has been called mentions a commit hash which includes that new commit. Was this intentional? I don't object to the new commit per se but it seems strange to be including it in a vote about committer policy.

File: policies/committer-policy.md (hunk 17 above)
Comment (mattcaswell) [reply policies/committer-policy.md:41]: I had the same doubts when we introduced the original review process - but I would not go back now. We always have the option of reverting this change if it doesn't work out well.

Comment (beldmit) [reply policies/committer-policy.md:41]: I agree with the original process, but I think this change will not increase the quality but just processing time. 
Could we establish some metrics?

Comment (mattcaswell) [reply policies/committer-policy.md:41]: Feasibly we could establish metrics for review time (e.g. measure time from PR Open to PR Close). I'm not sure how you would establish metrics for quality.

File: policies/committer-policy.md (hunk 18 above)
Comment (mspncp) [reply policies/committer-policy.md:51]: Only OMC/OTC members should be able to _force_ an OMC/OTC vote by placing a hold. Anyone else (committers or normal community members) still has the possibility to convince one or more OMC/OTC members of his opinion and get them to place the hold.

**Comment by mspncp** [comment]
> Question: This PR seems to have a new unrelated commit in it to define perlasm. Its the same commit as is present in #10. The vote that has been called mentions a commit hash which includes that new commit. Was this intentional? I don't object to the new commit per se but it seems strange to be including it in a vote about committer policy.

I don't see such a commit in this PR. Where did you notice it?

**Comment by mspncp** [comment]
Ah! I understand: Paul referred to the wrong commit id in the vote topic.

File: policies/committer-policy.md (hunk 17 above)
Comment (mspncp) [reply policies/committer-policy.md:41]: In some specific cases, it would be possible to improve the quality of the review by not just looking at the code but also download and build it locally, and then test it independently of the author. But I doubt that such an option could be turned into in a general requirement.

**Comment by paulidale** [comment]
It should have been 76bb566cf401e05bc54cd81ce429e83e2827278a.
I don't object to 3766d6ba2648e716e973af6e1821687ad46ee57c also going in, one less vote is good :)

**Comment by mattcaswell** [comment]
If the vote is actually for 76bb566 then I vote +1. If it is for 3766d6b then I vote 0 since I don't think its a good idea to be mixing votes like this, but I don't actively object to the content.

**Comment by kroeckx** [comment]
I'm voting +1

**Comment by paulidale** [comment]
I went with 76bb566cf401e05bc54cd81ce429e83e2827278a for simplicity

```
Topic: Accept the committer policy as of 76bb566cf401e05bc54cd81ce429e83e2827278a
Proposed by: Pauli
Issue link: https://github.com/openssl/general-policies/issues/8
Public: yes
Opened: 2022-02-17
Closed: 2022-02-21
Accepted:  yes  (for: 4, against: 0, abstained: 0, not voted: 2)

  Kurt       [+1]
  Mark       [  ]
  Matt       [+1]
  Pauli      [+1]
  Richard    [  ]
  Tim        [+1]
```

**Comment by levitte** [comment]
Vote: 0

**Comment by levitte** [comment]
(the label "ready to vote" was never set on this one...)

**Comment by iamamoose** [comment]
[0]
